## This exercise is meant to practice combining files in Python

### Import Libraries

In [7]:
import pandas as pd
import numpy as np
import os

### Create data to experiment on

In [3]:
# Define a dictionary containing data from January 2020

data1 = {'customer_id':['6732', '767', '890', '635'],
        'month':['Jan-20','Jan-20', 'Jan-20', 'Jan-20'],
        'purchased_meat':[0, 13, 3, 4],
        'purchase_alcohol':[1, 2, 10, 0],
        'purchased_snacks':[10, 5, 1, 7]}

In [4]:
# Define a dictionary containing data from February 2020

data2 = {'customer_id':['6732', '767', '890', '635'],
    'month':['Feb-20', 'Feb-20', 'Feb-20', 'Feb-20'],
    'purchased_meat':[0, 10, 5, 3],
    'purchased_alcohol':[2, 4, 14, 0],
    'purchased_snacks':[15, 3, 2, 6]}

### Convert dictionaries into dataframes

In [8]:
# Convert data1 and data2 into dataframes 

df_1 = pd.DataFrame(data1, index=[0, 1, 2, 3])
df_2 = pd.DataFrame(data2,index=[0, 1, 2, 3])

In [9]:
df_1

customer_id   month  purchased_meat  purchase_alcohol  purchased_snacks
0        6732  Jan-20               0                 1                10
1         767  Jan-20              13                 2                 5
2         890  Jan-20               3                10                 1
3         635  Jan-20               4                 0                 7

In [10]:
df_2

customer_id   month  purchased_meat  purchased_alcohol  purchased_snacks
0        6732  Feb-20               0                  2                15
1         767  Feb-20              10                  4                 3
2         890  Feb-20               5                 14                 2
3         635  Feb-20               3                  0                 6

### Concatenate newly created dataframes

In [11]:
# Create a lis - frames - containing the dataframes to be combined: df_1 and df_2

frames = [df_1,df_2]

# Concatenate

df_concat = pd.concat(frames)

In [12]:
df_concat

customer_id   month  purchased_meat  purchase_alcohol  purchased_snacks  \
0        6732  Jan-20               0               1.0                10   
1         767  Jan-20              13               2.0                 5   
2         890  Jan-20               3              10.0                 1   
3         635  Jan-20               4               0.0                 7   
0        6732  Feb-20               0               NaN                15   
1         767  Feb-20              10               NaN                 3   
2         890  Feb-20               5               NaN                 2   
3         635  Feb-20               3               NaN                 6   

   purchased_alcohol  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
0                2.0  
1                4.0  
2               14.0  
3                0.0

In [13]:
# Concatenate the dataframes using the axis = 1 --> create a wide format as an output

df_concat = pd.concat(frames, axis = 1)

In [14]:
# Check the output

df_concat

customer_id   month  purchased_meat  purchase_alcohol  purchased_snacks  \
0        6732  Jan-20               0                 1                10   
1         767  Jan-20              13                 2                 5   
2         890  Jan-20               3                10                 1   
3         635  Jan-20               4                 0                 7   

  customer_id   month  purchased_meat  purchased_alcohol  purchased_snacks  
0        6732  Feb-20               0                  2                15  
1         767  Feb-20              10                  4                 3  
2         890  Feb-20               5                 14                 2  
3         635  Feb-20               3                  0                 6

In [38]:
# Combining two dataframes using the df_merge() function

df_merged = df_1.merge(df_2, on = 'customer_id')

In [34]:
df_merged

customer_id month_x  purchased_meat_x  purchase_alcohol  purchased_snacks_x  \
0        6732  Jan-20                 0                 1                  10   
1         767  Jan-20                13                 2                   5   
2         890  Jan-20                 3                10                   1   
3         635  Jan-20                 4                 0                   7   

  month_y  purchased_meat_y  purchased_alcohol  purchased_snacks_y  
0  Feb-20                 0                  2                  15  
1  Feb-20                10                  4                   3  
2  Feb-20                 5                 14                   2  
3  Feb-20                 3                  0                   6

In [40]:
# Checking for full match

df_merged = df_1.merge(df_2, on = ['customer_id'], indicator = True)

In [41]:
df_merged

customer_id month_x  purchased_meat_x  purchase_alcohol  purchased_snacks_x  \
0        6732  Jan-20                 0                 1                  10   
1         767  Jan-20                13                 2                   5   
2         890  Jan-20                 3                10                   1   
3         635  Jan-20                 4                 0                   7   

  month_y  purchased_meat_y  purchased_alcohol  purchased_snacks_y _merge  
0  Feb-20                 0                  2                  15   both  
1  Feb-20                10                  4                   3   both  
2  Feb-20                 5                 14                   2   both  
3  Feb-20                 3                  0                   6   both

In [42]:
# Testing merge without overwriting data

pd.merge(df_1,df_2, on = ['customer_id'], indicator = True)

customer_id month_x  purchased_meat_x  purchase_alcohol  purchased_snacks_x  \
0        6732  Jan-20                 0                 1                  10   
1         767  Jan-20                13                 2                   5   
2         890  Jan-20                 3                10                   1   
3         635  Jan-20                 4                 0                   7   

  month_y  purchased_meat_y  purchased_alcohol  purchased_snacks_y _merge  
0  Feb-20                 0                  2                  15   both  
1  Feb-20                10                  4                   3   both  
2  Feb-20                 5                 14                   2   both  
3  Feb-20                 3                  0                   6   both

### Importing 'order_products_prior.csv as df_ords_prior

In [43]:
# Set path

path = r'/Users/dianaalatriste/Documents/Instacart'

In [45]:
# Import new dataframe df_ords_prior

df_ords_prior = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'order_products_prior.csv'), index_col = 0)

In [49]:
# Import df_ords dataframe

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = 0)

In [47]:
# Check output of df_ords_prior

df_ords_prior.head()

product_id  add_to_cart_order  reordered
order_id                                          
2              33120                  1          1
2              28985                  2          1
2               9327                  3          0
2              45918                  4          1
2              30035                  5          0

In [50]:
# Check output of df_ords

df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  last_purchase  
0                  8            NaN  
1                  7           15.0  
2                 12           21.0  
3                  7           29.0  
4                 15           28.0

In [51]:
# Check for shape of df_ords_prior

df_ords_prior.shape

(32434489, 3)

In [52]:
# Check for shape of df_ords

df_ords.shape

(3421083, 7)

In [53]:
# Merging df_ords_prior to df_ords without specifying type of merge - Therefore it uses defualt 'inner'
df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)

In [54]:
# Check output

df_merged_large.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2539329        1    prior             1                   2   
2   2539329        1    prior             1                   2   
3   2539329        1    prior             1                   2   
4   2539329        1    prior             1                   2   

   order_hour_of_day  last_purchase  product_id  add_to_cart_order  reordered  \
0                  8            NaN         196                  1          0   
1                  8            NaN       14084                  2          0   
2                  8            NaN       12427                  3          0   
3                  8            NaN       26088                  4          0   
4                  8            NaN       26405                  5          0   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

In [55]:
# Summing up values in merge column

df_merged_large['_merge'].value_counts()

both          32434489
left_only            0
right_only           0
Name: _merge, dtype: int64

In [57]:
# Check shape of df_merged_large

df_merged_large.shape

(32434489, 11)

#### We can see that there are only entries that have a value of “both,” leading you to think that your key column, “order_id,” exists completely in both dataframes. However, this conclusion is wrong. What pandas does here is fill in information about each product for every “order_id” in the df_ords dataframe, which is why the resulting dataframe has 32,434,489 rows (the same total count as the df_ords_prior dataframe).

#### The resulting table only contains observations found in both dataframes. As such, the merge flag here will only show entries that have a value of “both.” 

In [56]:
# Export data to pkl

df_merged_large.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.pkl'))

In [58]:
print('End of Exercise')

End of Exercise
